File: go/spec/services.go
package spec

import "cloud.google.com/go/civil"
import "github.com/google/uuid"
import "github.com/shopspring/decimal"

type EmptyDef struct{}

var Empty = EmptyDef{}

type EchoBodyResponse struct {
	Ok *Message
}

type EchoQueryResponse struct {
	Ok *Message
}

type EchoHeaderResponse struct {
	Ok *Message
}

type EchoUrlParamsResponse struct {
	Ok *Message
}

type IEchoService interface {
	EchoBody(body *Message) (*EchoBodyResponse, error)
	EchoQuery(intQuery int, stringQuery string) (*EchoQueryResponse, error)
	EchoHeader(intHeader int, stringHeader string) (*EchoHeaderResponse, error)
	EchoUrlParams(intUrl int, stringUrl string) (*EchoUrlParamsResponse, error)
}

type CheckQueryResponse struct {
	Ok *EmptyDef
}

type CheckUrlParamsResponse struct {
	Ok *EmptyDef
}

type CheckForbiddenResponse struct {
	Ok *Message
	Forbidden *EmptyDef
}

type ICheckService interface {
	CheckQuery(pString string, pStringOpt *string, pStringArray []string, pDate civil.Date, pDateArray []civil.Date, pDatetime civil.DateTime, pInt int, pLong int64, pDecimal decimal.Decimal, pEnum Choice, pStringDefaulted string) (*CheckQueryResponse, error)
	CheckUrlParams(intUrl int64, stringUrl string, floatUrl float32, boolUrl bool, uuidUrl uuid.UUID, decimalUrl decimal.Decimal, dateUrl civil.Date) (*CheckUrlParamsResponse, error)
	CheckForbidden() (*CheckForbiddenResponse, error)
}

File: ts-koa/src/spec/services.ts
import * as models from './models'

export interface EchoBodyParams {
    body: models.Message,
}

export type EchoBodyResponse =
    | { status: "ok", data: models.Message }

export interface EchoQueryParams {
    int_query: number,
    string_query: string,
}

export type EchoQueryResponse =
    | { status: "ok", data: models.Message }

export interface EchoHeaderParams {
    'int-header': number,
    'string-header': string,
}

export type EchoHeaderResponse =
    | { status: "ok", data: models.Message }

export interface EchoUrlParamsParams {
    int_url: number,
    string_url: string,
}

export type EchoUrlParamsResponse =
    | { status: "ok", data: models.Message }

export interface EchoService {
    echoBody(params: EchoBodyParams): Promise<EchoBodyResponse>
    echoQuery(params: EchoQueryParams): Promise<EchoQueryResponse>
    echoHeader(params: EchoHeaderParams): Promise<EchoHeaderResponse>
    echoUrlParams(params: EchoUrlParamsParams): Promise<EchoUrlParamsResponse>
}

export type CheckEmptyResponse =
    | { status: "ok" }

export interface CheckQueryParams {
    p_string: string,
    p_string_opt?: string | undefined,
    p_string_array: string[],
    p_date: string,
    p_date_array: string[],
    p_datetime: Date,
    p_int: number,
    p_long: number,
    p_decimal: number,
    p_enum: models.Choice,
    p_string_defaulted: string,
}

export type CheckQueryResponse =
    | { status: "ok" }

export interface CheckUrlParamsParams {
    int_url: number,
    string_url: string,
    float_url: number,
    bool_url: boolean,
    uuid_url: string,
    decimal_url: number,
    date_url: string,
}

export type CheckUrlParamsResponse =
    | { status: "ok" }

export type CheckForbiddenResponse =
    | { status: "ok", data: models.Message }
    | { status: "forbidden" }

export interface CheckService {
    checkEmpty(): Promise<CheckEmptyResponse>
    checkQuery(params: CheckQueryParams): Promise<CheckQueryResponse>
    checkUrlParams(params: CheckUrlParamsParams): Promise<CheckUrlParamsResponse>
    checkForbidden(): Promise<CheckForbiddenResponse>
}
